wdl/transforms/new-base/src/main/scala/wdl/transforms/base/wdlom2wom/expression/WdlomWomExpression.scala

```diff
@@ -36,9 +35,24 @@ final case class WdlomWomExpression private (expressionElement: ExpressionElemen
   // NB types can be determined using the linked values, so we don't need the inputMap:
   override def evaluateType(inputMap: Map[String, WomType]): ErrorOr[WomType] = evaluatedType
 
-  override def evaluateFiles(inputs: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[FileEvaluation]] =
-    // If WDL ever supports optional outputs then the `optional` parameter of `FileEvaluation` would need to be assigned appropriately.
-    expressionElement.evaluateFilesNeededToEvaluate(inputs, ioFunctionSet, coerceTo) map { _ map FileEvaluation.requiredFile }
+  /** Returns `true` if the specified `WomType` contains only optional file types. If the specified `WomType` is
+    * complex and we can't work out definitively whether all its file types are optional, return `false` to be on the
+    * safe side. */
+  def areAllFileTypesInWomTypeOptional(womType: WomType): Boolean = womType match {
+    case WomOptionalType(_: WomPrimitiveFileType) => true
+    case _: WomPrimitiveFileType => false
+    case _: WomPrimitiveType => true // WomPairTypes may have a non-File component here, this is fine.
+    case WomArrayType(inner) => areAllFileTypesInWomTypeOptional(inner)
+    case WomMapType(_, inner) => areAllFileTypesInWomTypeOptional(inner)
+    case WomPairType(leftType, rightType) => areAllFileTypesInWomTypeOptional(leftType) && areAllFileTypesInWomTypeOptional(rightType)
+    case _ => false
```

Comment:
We should have a case for structs (which I believe are expressed as `WomObject`s with defined shapes)

Reply:
in progress